docs/proposals/templates.md

```diff
@@ -0,0 +1,404 @@
+* As a developer I want to select the template to use from a list of available templates
+* Enabling the creation of user interfaces that can guide an application deployer through the deployment process with descriptive help about the configuration value decisions they are making, and useful default values where appropriate
+
+### Use cases for parameters within templates
+
+* Generate and share passwords between components (parameter value is provided to each component as an environment variable)
+* Allow for simple deployment-time customization of “app” configuration via environment values or api objects, e.g. memory 
+  tuning parameters to a MySQL image, Docker image registry prefix for image strings, pod resource requests and limits, default 
+  scale size.
+* Allow simple, declarative defaulting of parameter values AND expose them to end users in an approachable way - a parameter 
+  like “MySQL table space” can be parameterized in images as an env var - the template parameters declare the parameter, give 
+  it a friendly name, give it a reasonable default, and informs the user what tuning options available.
+* Customization of component names to avoid collisions and ensure matched labeling (eg replica selector value and pod label are 
+  user provided and in sync).
+* Customize cross-component references (eg user provides the name of a secret that already exists in their namespace, to use in 
+  a pod as a TLS cert).
+* Provide guidance to users for parameters such as default values, descriptions, and whether or not a particular parameter value 
+  is required or can be left blank.
+* Parameterize the replica count of a deployment or PetSet
+* Parameterize part of the labels and selector for a DaemonSet
+
+
+## Design Assumptions
+
+The goal for this proposal is a simple schema which addresses a few basic challenges:
+
+* Allow application authors to expose configuration knobs for application deployers, with suggested defaults and 
+descriptions of the purpose of each knob
+* Allow application deployers to easily customize exposed values like object names while maintaining referential integrity 
+  between dependent pieces
+* Support maintaining a library of templates within Kubernetes that can be accessed and instantiated by end users
+* Allow users to quickly and repeatedly deploy instances of well-defined application patterns produced by the community
+* Follow established Kubernetes API patterns by defining new template related APIs which consume+return first class Kubernetes 
+  API (and therefore json conformant) objects.
+
+We do not wish to invent a new turing complete templating language.  There are good options available 
+(e.g. https://github.com/mustache/mustache) for developers who want a completely flexible and powerful solution for creating 
+arbitrarily complex templates with parameters, and tooling can be built around such schemes.
+
```

Comment:
Up to this point, the document can read as a general proposal for templating in Kubernetes. I'd like to see it broken into two documents:
* A document that establishes the general requirements and design assumptions, consisting of the content up to this point, and
* A document proposing to use the OpenShift templating mechanism as the solution adopting to address those requirements.

The motivation for this request is to create a level playing field for the several technologies available to satisfy the general requirements. By splitting this document into two, all technology providers can collaborate on the requirements document, and each technology provider can submit a technology proposal.

I would further suggest that this effort be clearly advertised via the Config SIG mailing list and chat room, so that all technology providers can be apprised of its existence, and can choose to participate, if interested.